File: CPP/Lesson_049_min_stack.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 049 -- Min Stack
// Category   : Stack
// Difficulty : Medium
// Study Plan : Day 24
// =============================================================
//
// QUESTION:
//   Design a stack supporting push, pop, top, and getMin in O(1).
// =============================================================

#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
class MinStack {
    stack<int> s, m;
public:
    void push(int v){ s.push(v); m.push(m.empty()?v:min(v,m.top())); }
    void pop(){ s.pop(); m.pop(); }
    int top(){ return s.top(); }
    int getMin(){ return m.top(); }
};
int main(){ MinStack s; s.push(-2); s.push(0); s.push(-3); cout<<s.getMin()<<" "; s.pop(); cout<<s.top()<<" "<<s.getMin()<<endl; }

File: Go/Lesson_049_min_stack.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 049 -- Min Stack
// Category   : Stack
// Difficulty : Medium
// Study Plan : Day 24
// =============================================================
//
// QUESTION:
//   Design a stack supporting push, pop, top, and getMin in O(1).
// =============================================================

package main
import "fmt"
type MinStack struct{ s, m []int }
func (st *MinStack) Push(v int){
    st.s = append(st.s, v)
    if len(st.m)==0 || v < st.m[len(st.m)-1] { st.m = append(st.m, v) } else { st.m = append(st.m, st.m[len(st.m)-1]) }
}
func (st *MinStack) Pop(){ st.s = st.s[:len(st.s)-1]; st.m = st.m[:len(st.m)-1] }
func (st *MinStack) Top() int { return st.s[len(st.s)-1] }
func (st *MinStack) GetMin() int { return st.m[len(st.m)-1] }
func main(){
    s := &MinStack{}; s.Push(-2); s.Push(0); s.Push(-3)
    fmt.Println(s.GetMin()); s.Pop(); fmt.Println(s.Top()); fmt.Println(s.GetMin())
}
